admin/services/data-access/pin-generation.data.service.js

```diff
@@ -11,12 +11,48 @@ const serviceToExport = {
    * @return {Promise<*>}
    */
   sqlFindEligiblePupilsBySchool: async (schoolId, isLiveCheck) => {
-    const view = isLiveCheck === true ? 'vewPupilsEligibleForLivePinGeneration' : 'vewPupilsEligibleForFamiliarisationPinGeneration'
-    const sql = `SELECT 
-                  * 
-                FROM ${sqlService.adminSchema}.${view}
-                WHERE school_id=@schoolId                 
+    let sql
+    if (!isLiveCheck) {
+      sql = ` SELECT
+                  p.id,
+                  p.foreName,
+                  p.middleNames,
+                  p.lastName,
+                  p.dateOfBirth,
+                  p.school_id,
+                  g.group_id,
+                  p.urlSlug
+              FROM
+                [mtc_admin].[pupil] p
+                  LEFT JOIN [mtc_admin].[pupilAttendance] pa ON (p.id = pa.pupil_id and pa.isDeleted = 0)
+                  LEFT JOIN [mtc_admin].[attendanceCode] ac ON (pa.attendanceCode_id = ac.id )
+                  LEFT JOIN  [mtc_admin].[pupilGroup] g ON (g.pupil_id = p.id)
+                  INNER JOIN [mtc_admin].[pupilStatus] ps ON (p.pupilStatus_id = ps.id)
+              WHERE
+                -- include all pupils except those who are marked as not taking check because they left school
+                (ac.id IS NULL OR ac.code <> 'LEFTT')
+              AND    ps.code IN ('UNALLOC',
+                                'ALLOC',
+                                'LOGGED_IN')
+              AND school_id = ${schoolId}
+                  -- Exclude pupils who already have an active familiarisation check
+              AND p.id NOT IN (SELECT
```

Comment:
Actually, this could easily be a view by re-writing the sub-select:

```
  AND p.id NOT IN (SELECT
                       p2.id
                   FROM
                       [mtc_admin].[pupil] p2
                           LEFT JOIN [mtc_admin].[check] AS chk ON (p.id = chk.pupil_id)
                           LEFT JOIN [mtc_admin].[checkStatus] AS chkStatus ON (chk.checkStatus_id = chkStatus.id)
                   WHERE  chk.isLiveCheck = 0
                     AND    chkStatus.code IN ('NEW', 'STD', 'COL')
                     AND p2.school_id = p.school_id)`
```

Reply:
nice 👌  will try this out...

Reply:
It's only a part of the query though, what would be the advantage to do so?

Reply:
@GeorgeChatzigiannis its joining/binding via school id on the outer query, therefore less greedy on the `NOT IN` query due to the restriction.  When we call the view with `school_id=xxx` it should account for this in its plan and work over a reduced set, specific to the school in question.

Reply:
@GuyHarwood thanks for clarifying, that does make sense.